File: leetcode/2460_apply_operations_to_an_array.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
  public:
    static vector<int> applyOperations(const vector<int> &nums) noexcept {
        vector<int> ans{nums};
        applyOperations(ans);
        return ans;
    }

    static vector<int> applyOperations(vector<int> &nums) noexcept {
        vector<int> ans;
        ans.reserve(nums.size());

        for (int i{1}; i < nums.size() - 1; ++i) {
            if (nums[i] == nums[i - 1]) {
                nums[i - 1] *= 2;
                nums[i] = 0;
            }
        }

        for (const int v : ans)
            if (v)
                ans.push_back(v);

        while (ans.size() != nums.size())
            ans.push_back(0);

        return ans;
    }
};

int main() {
    Solution::applyOperations({1, 2, 2, 1, 1, 0}); // {1,4,2,0,0,0}
    Solution::applyOperations({0, 1});             // {1,0}
}

File: leetcode/2460_apply_operations_to_an_array.rs
struct Solution;

impl Solution {
    pub fn apply_operations(mut nums: Vec<i32>) -> Vec<i32> {
        let mut ans = Vec::new();

        for x in 1..nums.len() {
            nums[x - 1] = if nums[x] == nums[x - 1] {
                nums[x] = 0;
                nums[x - 1] * 2
            } else {
                nums[x - 1]
            };
        }

        for &x in nums.iter() {
            if x > 0 {
                ans.push(x);
            }
        }

        while ans.len() != nums.len() {
            ans.push(0);
        }

        ans
    }
}

fn main() {
    let t1 = Solution::apply_operations(vec![1, 2, 2, 1, 1, 0]);
    println!("{:?}", t1);
    let t2 = Solution::apply_operations(vec![0, 1]);
    println!("{:?}", t2);
}
